kupfer/plugin/id2/__init__.py

```diff
@@ -0,0 +1,195 @@
+__kupfer_actions__ = ("LookupID",)
+__description__ = _(
+    """Detect and open DOIs, ISBNs, OrcIDs, crypto-addresses, geo-coordinates, and more in your web-browser. Examples:
+
+- Unicode:  \t"U+1F194"
+- UnixTime: \t"1606324253"
+- Location: \t"8FWH4HX8+QR"
+- DOI:    \t\t"10.1109/FDL.2018.8524068"
+- OrcID:    \t"0000-0002-0006-7761"
+- BitCoin:  \t"3FZbgi29cpjq2GjdwV8eyHuJJnkLtktZc5"
+
+For a full list, see https://id2.dev
+
+To disable a certain ID, write a single character (e.g. "x") into the URL field below.
+"""
+)
+__version__ = "2022.1"
+__author__ = "Emanuel Regnath"
+
+
+from kupfer import utils, plugin_support
+from kupfer.objects import Action, TextLeaf, TextSource
+import re
+import json
+import sys
+import os
+import urllib.parse
+
+
+# global table holding identifier classes and icon names
+CLASS_NAMES = {
+    "i": ["Identifier", "edit-find"],
+    "g": ["Geo-Location", "applications-internet"],
+    "d": ["Document", "x-office-document"],
+    "o": ["Object", "package-x-generic"],
+    "h": ["Hash", "dialog-password"],
+    "t": ["Time", "x-office-calendar"],
+    "w": ["Person", "stock_person"],
+}
+
+# global data
+id2data = json.load(
+    open(os.path.join(os.path.dirname(__file__), "id2data.json"), "r"))
+
+
+# settings are auto-generated by "id2-settings-generator.py"
+__kupfer_settings__ = plugin_support.PluginSettings(
+    {
+        "key": "i_ip4",
+        "label": "IP4",
+        "type": str,
+        "value": "https://whois.domaintools.com/",
+    },
+    {
+        "key": "i_ip6",
+        "label": "IP6",
+        "type": str,
+        "value": "https://whois.domaintools.com/",
+    },
+    {
+        "key": "i_utf8",
+        "label": "UTF8",
+        "type": str,
+        "value": "https://www.compart.com/en/unicode/U+",
+    },
+    {
+        "key": "g_gps",
+        "label": "GPS",
+        "type": str,
+        "value": "https://www.google.com/maps/search/",
+    },
+    {
+        "key": "o_ean",
+        "label": "EAN",
+        "type": str,
+        "value": "https://www.ean-search.org/?q=",
+    },
+    {
+        "key": "o_upc",
+        "label": "UPC",
+        "type": str,
+        "value": "https://www.ean-search.org/?q=",
+    },
+    {"key": "o_asin", "label": "ASIN", "type": str, "value": "https://amzn.com/"},
+)
+
+
+# kupfer setting keys must not have colons
+def _settings_key(key: str):
+    return key.replace(":", "_")
+
+
+def _guess_id(token: str):
+    """Test if "token" matches the regex of any identifier and return all found types."""
+    token = token.strip()
+    types = []
+    for key, entry in id2data.items():
+        setkey = _settings_key(key)
+        if setkey in __kupfer_settings__ and len(__kupfer_settings__[setkey]) < 2:
+            continue
+        if len(token) in entry["lens"]:
+            match = re.match(r"^" + entry["re"] + r"$", token)
+            if match:
+                entry["part"] = match.group(1)
+                types.append(entry)
```

Comment:
better way - convert this function to generator and yield here entry; it's not necessary to create list and later iterate over it and yield value...